File: crates/hippocampus-java/src/test/java/io/github/lingtian303/hippocampus/HippocampusTest.java
package io.github.lingtian303.hippocampus;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Hippocampus Java 绑定的单元测试（v2.15-c 新增）。
 *
 * <p>共 12 个测试，覆盖版本/操作列表/创建/归档/检索/摘要/Prompt/周期任务/全链路，
 * 与 Go 绑定（{@code crates/hippocampus-go/hippocampus_test.go}）对齐。
 *
 * <p>运行测试前需确保 hippocampus 动态库已构建（{@code cargo build --release -p hippocampus-ffi}），
 * 且库路径已通过 pom.xml 的 surefire 插件配置（{@code jna.library.path}）注入。
 */
@DisplayName("Hippocampus Java 绑定测试")
class HippocampusTest {

    /** 临时目录（JUnit 5 自动创建和清理）。 */
    @TempDir
    Path tempDir;

    /**
     * 构造最小合法的 MessageTurn 数组 JSON（与 Go 测试保持一致）。
     *
     * <p>字段对应 Rust 端 {@code MessageTurn} 结构体：
     * id / user_message / llm_message / tags / timestamp / token_count。
     */
    private static String makeTurnsJson() {
        return "[{"
                + "\"id\":\"00000000-0000-0000-0000-000000000001\","
                + "\"user_message\":{"
                +   "\"text\":\"你好\","
                +   "\"attachments\":[],"
                +   "\"tool_calls\":[],"
                +   "\"thinking\":null"
                + "},"
                + "\"llm_message\":{"
                +   "\"text\":\"你好！有什么可以帮助你的吗？\","
                +   "\"attachments\":[],"
                +   "\"tool_calls\":[],"
                +   "\"thinking\":null"
                + "},"
                + "\"tags\":[{\"kind\":\"Text\"}],"
                + "\"timestamp\":\"2026-07-04T12:00:00Z\","
                + "\"token_count\":50"
                + "}]";
    }

    /** 构造临时存储路径。 */
    private String storagePath() {
        return tempDir.toString();
    }

    /* ====================================================================
     * 1. 常量与基础信息测试
     * ================================================================== */

    @Test
    @DisplayName("版本号应为 0.1.0")
    void testVersion() {
        System.out.println("版本号: " + Hippocampus.VERSION);
        assertEquals("0.1.0", Hippocampus.VERSION);
    }

    @Test
    @DisplayName("支持的操作列表应包含 5 个操作")
    void testOperations() {
        List<String> ops = Hippocampus.OPERATIONS;
        assertEquals(5, ops.size());
        assertTrue(ops.containsAll(Arrays.asList("archive", "retrieve", "summaries", "prompt", "compaction")));
    }

    /* ====================================================================
     * 2. 实例创建与生命周期测试
     * ================================================================== */

    @Test
    @DisplayName("创建实例（无 projectId）")
    void testNewHippocampus() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-1", null)) {
            assertNotNull(hp);
            System.out.println("创建成功: " + hp);
        }
    }

    @Test
    @DisplayName("创建实例（带 projectId）")
    void testNewHippocampusWithProjectId() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-2", "project-x")) {
            assertNotNull(hp);
        }
    }

    @Test
    @DisplayName("toString 应返回有效格式")
    void testHippocampusToString() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-3", null)) {
            assertEquals("Hippocampus(handle=valid)", hp.toString());
        }
    }

    @Test
    @DisplayName("close 后 toString 应显示 closed")
    void testHippocampusClose() {
        Hippocampus hp = Hippocampus.create(storagePath(), "session-4", null);
        hp.close();
        assertEquals("Hippocampus(closed)", hp.toString());
        // 多次 close 应安全（幂等）
        hp.close();
    }

    /* ====================================================================
     * 3. 错误处理测试
     * ================================================================== */

    @Test
    @DisplayName("archive 空 JSON 应抛出 IllegalArgumentException")
    void testArchiveEmptyJson() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-5", null)) {
            assertThrows(IllegalArgumentException.class, () -> hp.archive(""));
            assertThrows(IllegalArgumentException.class, () -> hp.archive("   "));
        }
    }

    @Test
    @DisplayName("archive 无效 JSON 应抛出 RuntimeException")
    void testArchiveInvalidJson() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-6", null)) {
            RuntimeException ex = assertThrows(RuntimeException.class, () -> hp.archive("invalid json"));
            System.out.println("预期错误: " + ex.getMessage());
            assertTrue(ex.getMessage().contains("解析") || ex.getMessage().toLowerCase().contains("parse"));
        }
    }

    @Test
    @DisplayName("retrieve 不存在的 hookId 应抛出 RuntimeException")
    void testRetrieveNonexistent() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-7", null)) {
            RuntimeException ex = assertThrows(RuntimeException.class,
                    () -> hp.retrieve("nonexistent-hook-id"));
            System.out.println("预期错误: " + ex.getMessage());
            assertTrue(ex.getMessage().contains("钩子") || ex.getMessage().toLowerCase().contains("hook"));
        }
    }

    @Test
    @DisplayName("summaries 空存储应返回 []")
    void testSummariesEmpty() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-8", null)) {
            String result = hp.summaries();
            assertEquals("[]", result);
        }
    }

    @Test
    @DisplayName("compaction 无效 period 应抛出 IllegalArgumentException")
    void testCompactionInvalidPeriod() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-9", null)) {
            IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                    () -> hp.compaction("daily"));
            System.out.println("预期错误: " + ex.getMessage());
            assertTrue(ex.getMessage().contains("period"));
        }
    }

    /* ====================================================================
     * 4. 端到端工作流测试
     * ================================================================== */

    @Test
    @DisplayName("完整工作流：归档 → 摘要 → Prompt → 检索")
    void testArchiveFullWorkflow() {
        try (Hippocampus hp = Hippocampus.create(storagePath(), "session-10", null)) {
            // 1. 归档
            String summaryJson = hp.archive(makeTurnsJson());
            assertNotNull(summaryJson);
            assertTrue(summaryJson.contains("hook_id"));
            System.out.println("归档成功，summary 包含 hook_id");

            // 从 summary 中提取 hook_id（简单字符串匹配，避免引入 JSON 库）
            String hookId = extractJsonValue(summaryJson, "hook_id");
            assertNotNull(hookId, "无法从 summary 提取 hook_id");
            System.out.println("hook_id: " + hookId);

            // 2. 摘要列表（应有 1 条）
            String summaries = hp.summaries();
            assertNotNull(summaries);
            assertTrue(summaries.contains(hookId));
            System.out.println("摘要列表包含已归档的 hook_id");

            // 3. 渲染 Prompt
            String prompt = hp.prompt();
            assertNotNull(prompt);
            System.out.println("渲染 prompt 长度: " + prompt.length() + " 字符");

            // 4. 检索完整记忆
            String memoryJson = hp.retrieve(hookId);
            assertNotNull(memoryJson);
            assertTrue(memoryJson.contains("turns") || memoryJson.contains("session"));
            System.out.println("检索成功，memory 包含 turns/session 字段");
        }
    }

    /** 从 JSON 字符串中提取指定字段的字符串值（简单实现，避免引入 JSON 库依赖）。 */
    private static String extractJsonValue(String json, String field) {
        String key = "\"" + field + "\":\"";
        int start = json.indexOf(key);
        if (start < 0) {
            return null;
        }
        start += key.length();
        int end = json.indexOf("\"", start);
        if (end < 0) {
            return null;
        }
        return json.substring(start, end);
    }
}

File: crates/hippocampus-go/hippocampus_test.go
// Hippocampus Go 绑定测试套件（v2.15）
//
// 12 个测试覆盖：版本/操作列表/构造/带项目构造/String/Close 幂等/
// 空 JSON/无效 JSON/不存在 hook_id/空 summaries/无效 period/端到端工作流。
//
// 与 Node 测试（crates/hippocampus-node/src/lib.rs）对齐，保证跨语言行为一致。
//
// 运行方式（Windows）：build.bat
// 运行方式（Linux/macOS）：make test

package hippocampus

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// ============================================================================
// 测试辅助
// ============================================================================

// tempStorage 创建临时存储目录，返回绝对路径。
//
// t.TempDir() 会在测试结束时自动清理，无需手动 defer。
func tempStorage(t *testing.T) string {
	t.Helper()
	dir := t.TempDir()
	storage := filepath.Join(dir, "data")
	if err := os.MkdirAll(storage, 0o755); err != nil {
		t.Fatalf("创建临时目录失败: %v", err)
	}
	return storage
}

// makeTurnsJSON 构造最小合法的 MessageTurn 数组 JSON。
//
// 与 Node 测试中的 turn 结构完全一致，保证跨语言测试数据相同。
func makeTurnsJSON() string {
	turns := []map[string]interface{}{{
		"id": "00000000-0000-0000-0000-000000000001",
		"user_message": map[string]interface{}{
			"text":        "你好",
			"attachments": []interface{}{},
			"tool_calls":  []interface{}{},
			"thinking":    nil,
		},
		"llm_message": map[string]interface{}{
			"text":        "你好！有什么可以帮你的？",
			"attachments": []interface{}{},
			"tool_calls":  []interface{}{},
			"thinking":    nil,
		},
		"tags":        []map[string]string{{"kind": "Text"}},
		"timestamp":   "2026-07-04T12:00:00Z",
		"token_count": 50,
	}}
	b, err := json.Marshal(turns)
	if err != nil {
		panic("构造测试 turns JSON 失败: " + err.Error())
	}
	return string(b)
}

// ============================================================================
// 模块级测试
// ============================================================================

// TestVersion 验证版本号非空。
func TestVersion(t *testing.T) {
	if Version == "" {
		t.Fatal("Version 不应为空")
	}
	t.Logf("版本号: %s", Version)
}

// TestOperations 验证支持的操作列表包含 5 个核心操作。
func TestOperations(t *testing.T) {
	ops := Operations()
	if len(ops) != 5 {
		t.Fatalf("应有 5 个操作，实际 %d: %v", len(ops), ops)
	}
	expected := []string{"archive", "retrieve", "summaries", "prompt", "compaction"}
	for _, exp := range expected {
		found := false
		for _, op := range ops {
			if op == exp {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("操作列表缺少 %q，实际: %v", exp, ops)
		}
	}
}

// ============================================================================
// 构造与生命周期测试
// ============================================================================

// TestNewHippocampus 验证创建实例成功。
func TestNewHippocampus(t *testing.T) {
	storage := tempStorage(t)
	hp, err := New(storage, "sess-1", "")
	if err != nil {
		t.Fatalf("创建 Hippocampus 失败: %v", err)
	}
	defer hp.Close()

	if hp.handle == nil {
		t.Error("handle 不应为 nil")
	}
}

// TestNewHippocampusWithProjectID 验证带 project_id 创建实例。
func TestNewHippocampusWithProjectID(t *testing.T) {
	storage := tempStorage(t)
	hp, err := New(storage, "sess-proj", "proj-a")
	if err != nil {
		t.Fatalf("创建失败: %v", err)
	}
	defer hp.Close()
}

// TestHippocampusString 验证 String() 包含 "Hippocampus"。
func TestHippocampusString(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-str", "proj-x")
	defer hp.Close()

	s := hp.String()
	if !strings.Contains(s, "Hippocampus") {
		t.Errorf("String 应包含 Hippocampus，实际: %q", s)
	}
}

// TestHippocampusClose 验证 Close 幂等（多次调用不 panic）。
func TestHippocampusClose(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-close", "")

	// 多次 Close 应安全
	hp.Close()
	hp.Close()

	// Close 后 String 应返回 closed 状态
	s := hp.String()
	if !strings.Contains(s, "closed") {
		t.Errorf("Close 后 String 应标识 closed，实际: %q", s)
	}
}

// ============================================================================
// Archive 错误处理测试
// ============================================================================

// TestArchiveEmptyJSON 验证空 turnsJSON 返回错误。
func TestArchiveEmptyJSON(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-empty", "")
	defer hp.Close()

	// 空字符串应在 Go 层拦截（不调用 FFI）
	_, err := hp.Archive("")
	if err == nil {
		t.Error("空 turnsJSON 应返回错误")
	}

	// 空白字符串也应被拦截
	_, err = hp.Archive("   ")
	if err == nil {
		t.Error("空白 turnsJSON 应返回错误")
	}
}

// TestArchiveInvalidJSON 验证无效 JSON 返回错误（FFI 层反序列化失败）。
func TestArchiveInvalidJSON(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-invalid", "")
	defer hp.Close()

	_, err := hp.Archive("not a json")
	if err == nil {
		t.Error("无效 JSON 应返回错误")
	}
	t.Logf("预期错误: %v", err)
}

// ============================================================================
// Retrieve 错误处理测试
// ============================================================================

// TestRetrieveNonexistent 验证检索不存在的 hook_id 返回错误。
func TestRetrieveNonexistent(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-retrieve", "")
	defer hp.Close()

	_, err := hp.Retrieve("nonexistent-hook-id")
	if err == nil {
		t.Error("检索不存在的 hook_id 应返回错误")
	}
	t.Logf("预期错误: %v", err)
}

// ============================================================================
// Summaries 测试
// ============================================================================

// TestSummariesEmpty 验证空存储返回 "[]"（非错误）。
func TestSummariesEmpty(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-summ", "")
	defer hp.Close()

	got, err := hp.Summaries()
	if err != nil {
		t.Fatalf("Summaries 失败: %v", err)
	}
	if strings.TrimSpace(got) != "[]" {
		t.Errorf("空存储应返回 []，实际: %q", got)
	}
}

// ============================================================================
// Compaction 错误处理测试
// ============================================================================

// TestCompactionInvalidPeriod 验证无效 period 返回错误（Go 层拦截，不调用 FFI）。
func TestCompactionInvalidPeriod(t *testing.T) {
	storage := tempStorage(t)
	hp, _ := New(storage, "sess-comp", "")
	defer hp.Close()

	_, err := hp.Compaction("daily")
	if err == nil {
		t.Error("无效 period 应返回错误")
	}
	t.Logf("预期错误: %v", err)
}

// ============================================================================
// 端到端工作流测试
// ============================================================================

// TestArchiveFullWorkflow 端到端验证 archive → retrieve → summaries → prompt。
//
// 与 Node 测试 test_hippocampus_archive_full_workflow 对齐。
func TestArchiveFullWorkflow(t *testing.T) {
	storage := tempStorage(t)
	hp, err := New(storage, "sess-e2e", "")
	if err != nil {
		t.Fatalf("创建失败: %v", err)
	}
	defer hp.Close()

	// 1. archive
	summaryJSON, err := hp.Archive(makeTurnsJSON())
	if err != nil {
		t.Fatalf("Archive 失败: %v", err)
	}
	var summary map[string]interface{}
	if err := json.Unmarshal([]byte(summaryJSON), &summary); err != nil {
		t.Fatalf("解析 summary JSON 失败: %v\n原始 JSON: %s", err, summaryJSON)
	}
	hookID, ok := summary["hook_id"].(string)
	if !ok || hookID == "" {
		t.Fatalf("summary 缺少 hook_id 或类型错误: %v", summary)
	}
	t.Logf("归档成功，hook_id: %s", hookID)

	// 2. retrieve
	memoryJSON, err := hp.Retrieve(hookID)
	if err != nil {
		t.Fatalf("Retrieve 失败: %v", err)
	}
	var memory map[string]interface{}
	if err := json.Unmarshal([]byte(memoryJSON), &memory); err != nil {
		t.Fatalf("解析 memory JSON 失败: %v", err)
	}
	if memory["session_id"] != "sess-e2e" {
		t.Errorf("session_id 应为 sess-e2e，实际: %v", memory["session_id"])
	}
	turns, ok := memory["turns"].([]interface{})
	if !ok {
		t.Fatalf("memory.turns 应为数组，实际: %T", memory["turns"])
	}
	if len(turns) != 1 {
		t.Errorf("应有 1 个 turn，实际 %d", len(turns))
	}

	// 3. summaries
	summariesJSON, err := hp.Summaries()
	if err != nil {
		t.Fatalf("Summaries 失败: %v", err)
	}
	var summaries []map[string]interface{}
	if err := json.Unmarshal([]byte(summariesJSON), &summaries); err != nil {
		t.Fatalf("解析 summaries JSON 失败: %v", err)
	}
	if len(summaries) != 1 {
		t.Errorf("应有 1 条摘要，实际 %d", len(summaries))
	}

	// 4. prompt
	prompt, err := hp.Prompt()
	if err != nil {
		t.Fatalf("Prompt 失败: %v", err)
	}
	if prompt == "" {
		t.Error("prompt 不应为空（已归档 1 条记忆）")
	}
	t.Logf("渲染 prompt 长度: %d 字符", len(prompt))
}
